File: SampleProjects/GoT-Characters-master/Objective-C/GoT/Article.m
//
//  Article.m
//  GoT
//
//  Created by Paciej on 25/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

#import "Article.h"

static NSString *kArticleIdentifier = @"privateIdentifier";
static NSString *kArticleTitle = @"privateTitle";
static NSString *kArticleAbstract = @"privateAbstract";
static NSString *kArticleThumbnailURL = @"privateThumbnailURLString";
static NSString *kArticleThumbnailData = @"thumbnailData";

@interface Article ()

@property(nonatomic, copy, nonnull) NSString *privateIdentifier;
@property(nonatomic, copy, nonnull) NSString *privateTitle;
@property(nonatomic, copy, nonnull) NSString *privateAbstract;
@property(nonatomic, copy, nonnull) NSString *privateThumbnailURLString;
@property(nonatomic, copy, nonnull) NSString *privateUrlString;

@end

@implementation Article

- (nonnull instancetype)initWithIdentifier:(nonnull NSString *)identifier
                                     title:(nonnull NSString *)title
                                  abstract:(nonnull NSString *)abstract
                                 urlString:(nonnull NSString *)urlString
                        thumbnailURLString:
(nonnull NSString *)thumbnailURLString {
    self = [super init];
    if (self) {
        self.privateIdentifier = identifier;
        self.privateTitle = title;
        self.privateAbstract = abstract;
        self.privateUrlString = urlString;
        self.privateThumbnailURLString = thumbnailURLString;
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)aCoder {
    [aCoder encodeObject:self.privateIdentifier forKey:kArticleIdentifier];
    [aCoder encodeObject:self.privateTitle forKey:kArticleTitle];
    [aCoder encodeObject:self.privateAbstract forKey:kArticleAbstract];
    [aCoder encodeObject:self.privateThumbnailURLString
                  forKey:kArticleThumbnailURL];
    [aCoder encodeObject:self.thumbnailData forKey:kArticleThumbnailData];
}

- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super init];
    if (self) {
        self.privateIdentifier = [aDecoder decodeObjectForKey:kArticleIdentifier];
        self.privateTitle = [aDecoder decodeObjectForKey:kArticleTitle];
        self.privateAbstract = [aDecoder decodeObjectForKey:kArticleAbstract];
        self.privateThumbnailURLString =
        [aDecoder decodeObjectForKey:kArticleThumbnailURL];
        self.thumbnailData = [aDecoder decodeObjectForKey:kArticleThumbnailData];
    }
    return self;
}

- (nullable UIImage *)imageFromThumbnailData {
    return [UIImage imageWithData:self.thumbnailData];
}

- (NSString *)identifier {
    return self.privateIdentifier;
}

- (NSString *)title {
    return self.privateTitle;
}

- (NSString *)abstract {
    return self.privateAbstract;
}

- (NSString *)thumbnailURLString {
    return self.privateThumbnailURLString;
}

- (NSString *)urlString {
    return self.privateUrlString;
}
@end

File: SampleProjects/GoT-Characters-master/Swift/GoT/Article.swift
//
//  Article.swift
//  GoT
//
//  Created by Paciej on 25/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

import UIKit

private var kArticleIdentifier = "privateIdentifier"
private var kArticleTitle = "privateTitle"
private var kArticleAbstract = "privateAbstract"
private var kArticleThumbnailURL = "privateThumbnailURLString"
private var kArticleThumbnailData = "thumbnailData"

class Article: NSObject, NSCoding {
    var title: String {
        return privateTitle
    }
    var identifier: String {
        return privateIdentifier
    }
    var abstract: String {
        return privateAbstract
    }
    var urlString: String {
        return privateUrlString
    }
    var thumbnailURLString: String {
        return privateThumbnailURLString
    }
    var thumbnailData: Data?

    private var privateIdentifier = ""
    private var privateTitle = ""
    private var privateAbstract = ""
    private var privateThumbnailURLString = ""
    private var privateUrlString = ""

    init(identifier: String, title: String, abstract: String, urlString: String, thumbnailURLString: String) {
        super.init()
        
        privateIdentifier = identifier
        privateTitle = title
        privateAbstract = abstract
        privateUrlString = urlString
        privateThumbnailURLString = thumbnailURLString
    
    }

    func imageFromThumbnailData() -> UIImage? {
        if let aData = thumbnailData {
            return UIImage(data: aData)
        }
        return nil
    }

    func encode(with aCoder: NSCoder) {
        aCoder.encode(privateIdentifier, forKey: kArticleIdentifier)
        aCoder.encode(privateTitle, forKey: kArticleTitle)
        aCoder.encode(privateAbstract, forKey: kArticleAbstract)
        aCoder.encode(privateThumbnailURLString, forKey: kArticleThumbnailURL)
        aCoder.encode(thumbnailData, forKey: kArticleThumbnailData)
    }

    required init?(coder aDecoder: NSCoder) {
        super.init()
        
        privateIdentifier = aDecoder.decodeObject(forKey: kArticleIdentifier) as? String ?? ""
        privateTitle = aDecoder.decodeObject(forKey: kArticleTitle) as? String ?? ""
        privateAbstract = aDecoder.decodeObject(forKey: kArticleAbstract) as? String ?? ""
        privateThumbnailURLString = aDecoder.decodeObject(forKey: kArticleThumbnailURL) as? String ?? ""
        thumbnailData = aDecoder.decodeObject(forKey: kArticleThumbnailData) as? Data
    
    }
}